### PR TITLE
Adds a form to set and update Name and Affiliation, to be used when authoring circulars

### PR DESCRIPTION
**Pull request by dakota002**
An option for saving username and association which we might need for circulars.  

Relates to https://github.com/nasa-gcn/gcn.nasa.gov/issues/291

image

### CODE REVIEW

File: app.arc [hunk 1]
```diff
@@ -25,6 +25,9 @@ email_notification_subscription
   uuid *String
   topic **String
 
+user_data
```
Comment (lpsinger) [inline app.arc:28]:
```suggestion
user
```

File: app/routes/user/index.tsx [hunk 2]
```diff
@@ -7,21 +7,68 @@
...
 export default function User() {
-  const { email, idp } = useLoaderData<typeof loader>()
+  const { email, idp, username, association } = useLoaderData<typeof loader>()
   return (
     <>
       <h1>Welcome, {email}!</h1>
       <p>You signed in with {idp || 'username and password'}.</p>
+      <h2>Profile Info</h2>
+      <Form method="post">
+        <Label htmlFor="username">Display Name</Label>
```
Comment (lpsinger) [inline app/routes/user/index.tsx:52]:
```suggestion
        <Label htmlFor="username">Display Name</Label>
```

Add some explanatory text to the effect of: "How would you like your name to appear in GCN Circulars? For example: A. E. Einstein, A. Einstein, Albert Einstein"

File: app/routes/user/index.tsx [hunk 3]
```diff
@@ -7,21 +7,68 @@
...
+        <TextInput
+          id="username"
+          name="username"
+          type="text"
+          defaultValue={username}
+          required
+        />
+        <Label htmlFor="association">Association</Label>
```
Comment (lpsinger) [inline app/routes/user/index.tsx:60]: Let's call this Affiliation and give some examples. Include example affiliations for universities, national labs, private companies, and amateur astronomers.

File: app/routes/user/user_data.server.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    if (!user) throw new Response('not signed in', { status: 403 })
+    return new this(user.sub)
+  }
+
+  async updateUserData(username: string, association: string) {
+    const db = await tables()
+
+    await db.user_data.update({
+      Key: { sub: this.#sub },
+      UpdateExpression: 'set username = :username, association = :association',
```
Comment (lpsinger) [inline app/routes/user/user_data.server.ts:30]: There are too many things that might be considered a "username". Let's call this field "name" or "displayName".

Also... would be good to populate the default value from the user's OpenID scope.

**Review by lpsinger — CHANGES_REQUESTED** [review]
Change "Association" to "Affiliation".

File: app/routes/user/index.tsx (hunk 3 above)
Comment (dakota002) [reply app/routes/user/index.tsx:60]: Updating this now, is there an appropriate or specific way that amateur astronomers should identify? This is what I have at the moment, after skimming through some existing circulars:

image

File: app/routes/user/index.tsx (hunk 2 above)
Comment (dakota002) [reply app/routes/user/index.tsx:52]: Good call, will have this updated in the next commit

File: app/routes/user/index.tsx (hunk 3 above)
Comment (lpsinger) [reply app/routes/user/index.tsx:60]: @jracusin?

File: app/routes/user/user_data.server.ts (hunk 4 above)
Comment (dakota002) [reply app/routes/user/user_data.server.ts:30]: Do we have that available in scope info? Or would we need to update the mappings in cognito?

Comment (lpsinger) [reply app/routes/user/user_data.server.ts:30]: `openid` is a standard scope. Refer to the OpenID Connect documentation for the standard claims included in that scope.

Or, `console.log` the ID token.

File: app/routes/user/index.tsx (hunk 3 above)
Comment (lpsinger) [reply app/routes/user/index.tsx]: Ping, @jracusin.

File: app/routes/user/index.tsx [hunk 5]
```diff
@@ -7,21 +7,77 @@
...
+import { getFormDataString } from '~/lib/utils'
 import { getUser } from '../__auth/user.server'
+import { UserDataServer } from './user_data.server'
 
 export async function loader({ request }: DataFunctionArgs) {
   const user = await getUser(request)
   if (!user) throw new Response(null, { status: 403 })
-  return { email: user.email, idp: user.idp }
+  const server = await UserDataServer.create(request)
+  const addlUserData = await server.getUserData()
```
Comment (lpsinger) [inline app/routes/user/index.tsx:21]: addl?

File: app/routes/user/index.tsx [hunk 6]
```diff
@@ -7,21 +7,77 @@
...
+          id="displayName"
+          name="displayName"
+          type="text"
+          defaultValue={displayName}
+          required
+        />
+        <Label htmlFor="affiliation">Affiliation</Label>
+        <small className="text-base">
+          For example: Pennsylvania State University, Ioffe Institute, DESY,
+          Fermi-GBM Team, or amateur astronomers
```
Comment (lpsinger) [inline app/routes/user/index.tsx:68]:
```suggestion
          Fermi-GBM Team, or AAVSO
```

**Review by lpsinger — CHANGES_REQUESTED** [review]
What do we have in the standard openid claims that might fulfill a reasonable default display name?

File: app/routes/user/index.tsx (hunk 5 above)
Comment (dakota002) [reply app/routes/user/index.tsx:21]: "additional" abbreviated

File: app/routes/user/index.tsx [hunk 7]
```diff
@@ -7,21 +7,84 @@
...
+import { UserDataServer } from './user_data.server'
 
 export async function loader({ request }: DataFunctionArgs) {
   const user = await getUser(request)
   if (!user) throw new Response(null, { status: 403 })
-  return { email: user.email, idp: user.idp }
+  return {
+    email: user.email,
+    idp: user.idp,
+    displayName: user.name,
```
Comment (lpsinger) [inline app/routes/user/index.tsx:23]:
```suggestion
    name: user.name,
```

File: app/routes/user/user_data.server.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    this.#cookie = cookie
+  }
+
+  static async create(request: Request) {
+    const user = await getUser(request)
+    if (!user) throw new Response('not signed in', { status: 403 })
+    return new this(user.sub, request.headers.get('Cookie') ?? '')
+  }
+
+  async updateUserData(displayName: string, affiliation: string) {
```
Comment (lpsinger) [inline app/routes/user/user_data.server.ts:31]:
```suggestion
  async updateUserData(name: string, affiliation: string) {
```

File: app/routes/user/user_data.server.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    const session = await storage.getSession(this.#cookie)
+    const client = new CognitoIdentityProviderClient({})
+    const command = new UpdateUserAttributesCommand({
+      UserAttributes: [
+        {
+          Name: 'name',
+          Value: displayName,
```
Comment (lpsinger) [inline app/routes/user/user_data.server.ts:38]:
```suggestion
          Value: name,
```

**Review by lpsinger — CHANGES_REQUESTED** [review]
General comments:

- Change `displayName` to `name`
- Tolerate name and affiliation being absent from the token

File: app/routes/__auth/user.server.ts [hunk 10]
```diff
@@ -52,6 +54,8 @@ export function parseTokenSet(tokenSet: TokenSet): {
   const groups = ((claims['cognito:groups'] ?? []) as string[]).filter(
     (group) => group.startsWith('gcn.nasa.gov/')
   )
+  const name = claims.name
+  const affiliation = (claims['custom:affiliation'] ?? '') as string
```
[the remark this answers is not in the document]
Comment (dakota002) [reply app/routes/__auth/user.server.ts:58]: This seems to resolve issues with allowing the field to be blank, and the added one to the name field also helps with this

[the remark this answers is not in the document]
Comment (lpsinger) [reply app/routes/__auth/user.server.ts:58]: Yes, but it means that using the `?.` operator won't work.

**Review by lpsinger — APPROVED** [review]

**Pull request merged**